irr/src/OpenGL/VertexArrayRef.cpp

```diff
@@ -0,0 +1,165 @@
+#include "OpenGL/VertexArrayRef.h"
+#include "OpenGL/VertexType.h"
+#include "vendor/gl.h"
+#include "mt_opengl.h"
+#include "os.h"
+
+namespace irr
+{
+namespace scene
+{
+
+inline GLenum to_gl_usage(E_HARDWARE_MAPPING mapping)
+{
+	GLenum usage = GL_STATIC_DRAW;
+	if (mapping == scene::EHM_STREAM)
+		usage = GL_STREAM_DRAW;
+	else if (mapping == scene::EHM_DYNAMIC)
+		usage = GL_DYNAMIC_DRAW;
+
+	return usage;
+}
+
+inline GLenum to_gl_indextype(video::E_INDEX_TYPE index_type)
+{
+	GLenum indexType = 0;
+
+	switch (index_type) {
+    case (video::EIT_16BIT): {
+		indexType = GL_UNSIGNED_SHORT;
+		break;
+	}
+    case (video::EIT_32BIT): {
+		indexType = GL_UNSIGNED_INT;
+		break;
+	}
+	}
+
+	return indexType;
+}
+
+COpenGL3VertexArrayRef::COpenGL3VertexArrayRef()
+	: IVertexArrayRef()
+{
+	GL.GenVertexArrays(1, &VertexArrayId);
+	GL.GenBuffers(1, &VertexBufferId);
```

Comment:
mixing GL calls into the buffer code is very wrong because it means only the thread that is holding he GL context can create buffers.

it looks like you're also undoing the separation of vertex and index buffers I added recently

Reply:
My PR does the separation of the vertex buffer data management and the actual buffer (which is just a reference to the GPU one). The current implementation used through all the client part (`IMeshBuffer`) mixes all together into one place which is not very a good approach, gives a less control over the buffer state and much complicates the integration of the VAO which is the one of the main goals of the PR. So, this VAO reference object can be created only in the thread owning the GL context, the vertex and index data can be created anywhere without the duplicating to the buffer which is a good optimisation too decreasing the RAM usage.

Reply:
And yes, I will also try to return back the vertex and index buffer separation since I have not finished the rebase yet.